### PR TITLE
Block Verification Stubs

### PR DESCRIPTION
**Pull request by teor2345**
Stubs for context-free (zebra-chain) and context-dependent (zebra-consensus) verification of blocks and transactions.

TODO:
- [x] write tests
- [x] squash the fixups and 07ef270e45242b30ff30f1d987692853912b74d7

### CODE REVIEW

**Comment by teor2345** [comment]
This PR is the start of the verification work in #428. It might also cover some of the checkpointing in #429.

**Comment by dconnolly** [comment]
@teor2345 I think the 'project assign' job is failing because you're on a fork, you should have permission to create branches on the main repo if you want.

**Comment by teor2345** [comment]
> @teor2345 I think the 'project assign' job is failing because you're on a fork, you should have permission to create branches on the main repo if you want.

I'll try to remember to do that next time. But let's keep #449 open to fix the issue for external contributors.

File: zebra-chain/src/verify.rs
Comment (yaahc) [inline zebra-chain/src/verify.rs]: associated data?

File: zebra-consensus/src/verify.rs
Comment (yaahc) [inline zebra-consensus/src/verify.rs]: this can be removed I think
```suggestion
```

File: zebra-chain/src/verify.rs
Comment (yaahc) [inline zebra-chain/src/verify.rs]: this can be removed I think
```suggestion
```

File: zebra-consensus/src/verify.rs
Comment (yaahc) [inline zebra-consensus/src/verify.rs]: Are these verify stubs implicitly also AddBlocks for the chain state? We might want to use different terminology here, though I don't think that should be a blocker for merging this PR.

File: zebra-consensus/src/verify.rs
Comment (yaahc) [inline zebra-consensus/src/verify.rs]: This seems strange to me. I would expect a future that is waiting for more information to return a `Poll::pending`, not a `Poll::Ready(Response::PendingSomething)`. It seems like this is creating a future like interface on top of a service where we're using multiple request / responses to get the state of the _actual_ future we're representing behind the scenes.

This is fine for a first pass stub I think but I feel like we should review this more carefully to be sure this is the best approach, or at least so I can understand why this is the best approach.

**Review by yaahc — APPROVED** [review]
This looks good, particularly the documentation which is straight up fantastic. Just some minor lints and requests for clarification that don't need to block this PR imo.

File: zebra-consensus/src/verify.rs
Comment (teor2345) [reply zebra-consensus/src/verify.rs]: Thanks, I asked some questions in slack that will help me refine this design.

File: zebra-chain/src/verify.rs
Comment (teor2345) [reply zebra-chain/src/verify.rs]: The `block` and `transaction` items in the struct-like enum variants.

File: zebra-consensus/src/verify.rs
Comment (teor2345) [reply zebra-consensus/src/verify.rs]: I think I might just call it `AddBlock` ?
`VerifyAndAddBlock` is another option.

File: zebra-chain/src/verify.rs
Comment (teor2345) [reply zebra-chain/src/verify.rs]: When I try, the `!#[deny(missing_docs)]` module-level lint triggers on the `block` and `transaction` items in the struct-like enum variants.

File: zebra-consensus/src/verify.rs
Comment (teor2345) [reply zebra-consensus/src/verify.rs]: I want to add `!#[deny(missing_docs)]` to this module. Then I think this allow will be required.

File: zebra-chain/src/verify.rs
Comment (yaahc) [reply zebra-chain/src/verify.rs:19]: ah, agreed, though I don't think the decision not to document the data members needs to be documented. Though if you want to leave this in I think that's fine, though I'd change the comment to not make it sound like you're talking about associated types on traits.

```suggestion
// The data members of the enum variants don't need separate documentation.
```

File: zebra-consensus/src/verify.rs
Comment (yaahc) [reply zebra-consensus/src/verify.rs]: ooh, I get this now. I didn't realize that it would try to force you to document every data member as well, this all makes a lot more sense. I think we should just update the wording to be a little more clear, `associated data` makes me think traits and associated types.

File: zebra-consensus/src/verify.rs
Comment (yaahc) [reply zebra-consensus/src/verify.rs]: I think `AddBlock` makes the most sense, the verify seems implied by the service the AddBlock is being called on.

**Comment by hdevalence** [comment]
Thanks for getting the ball rolling on this! A few comments/questions:

1. Currently this PR (and the preexisting `design.md` document) puts the semantic validation code in `zebra-chain`.  In the meantime since writing the original `design.md`, we came up with the idea for the futures-based verification design, and on second thought I think that it would be cleaner if we put *all* of the verification logic in `zebra-consensus`, and just left `zebra-chain` as the data structure definitions.  A few upsides to this are that all of the validation logic is in one place, all of the tower-using code is in one place, `zebra-chain` doesn't need to have a Tokio dependency, etc.

2. The current verification stubs are aiming to handle both verification of the block chain and verification of mempool transactions.  But these are pretty different cases and I worry that they cause false sharing in the design.  The mempool is really only useful to create blocks as a miner (which we're not aiming to support initially) or to gossip transactions to other peers so that miners will eventually get them.  The second part is still important to contribute to the health of the network, but it's a separate goal from validating the chain.

3. The update to `design.md` notes that the `zebra-consensus` crate is responsible for validating blocks and mempool entries (and updating the respective states).  I think that this is (conceptually) the entire functionality, so that it would be enough for the `zebra-consensus` crate to expose just a block verifier and a mempool type, and have everything else be an internal implementation detail. For instance, blocks are committed to the chain atomically, so the only reason to verify a transaction is in the context of a block verification or when adding the transaction to the mempool.  But both of those cases can be handled internally to a service call, so they don't need to be exported.  And rather than having one service that verifies blocks or verifies transactions, we could have multiple services, only some of which are exported.  Combining this with point (1) gives a huge savings in exported API surface, because rather than having to export all the details of transaction, proof, etc. verification as part of the `zebra-chain` API and then consume it here, we can keep everything encapsulated into `zebra-consensus`.

4. If we only expose block verification or mempool updates, and these two chunks of functionality are separate, we may be able to simplify the `Request` type, albeit with a slight loss of functionality.  Currently the `Request` type is either `VerifyBlock`, `VerifyTransaction`, or `Flush`.  If the block verification service is only processing blocks, we can drop the `VerifyTransaction` enum, and if we handle the batching internally to the `Service`, we can drop the `Flush` variant, in which case we wouldn't need a `Request` enum at all, and just have `Service<Block, Response=BlockHeaderHash, Error=Report>` or similar.  The uncertain part is how to handle the batching internally, which we could perhaps do with a timeout layer or some other mechanism. I'm a little less sure on this point than on the others, but it seems much cleaner to me, so for that reason it seems worthwhile to try to pursue it first.

5. (edit) Also, if we build the block verification service out of a bunch of inner services then the `Pending` responses kind of naturally turn into `.await` points of an async block inside the outer `Service` call.

**Comment by dconnolly** [comment]
> I think that it would be cleaner if we put _all_ of the verification logic in `zebra-consensus`, and just left `zebra-chain` as the data structure definitions.  A few upsides to this are that all of the validation logic is in one place, all of the tower-using code is in one place, `zebra-chain` doesn't need to have a Tokio dependency, etc.
👍

> The mempool is really only useful to create blocks as a miner (which we're not aiming to support initially) or to gossip transactions to other peers so that miners will eventually get them.  The second part is still important to contribute to the health of the network, but it's a separate goal from validating the chain.
👍

> t it would be enough for the `zebra-consensus` crate to expose just a block verifier and a mempool type, and have everything else be an internal implementation detail.
👍

> For instance, blocks are committed to the chain atomically, so the only reason to verify a transaction is in the context of a block verification or when adding the transaction to the mempool.  But both of those cases can be handled internally to a service call, so they don't need to be exported.  And rather than having one service that verifies blocks or verifies transactions, we could have multiple services, only some of which are exported.  Combining this with point (1) gives a huge savings in exported API surface, because rather than having to export all the details of transaction, proof, etc. verification as part of the `zebra-chain` API and then consume it here, we can keep everything encapsulated into `zebra-consensus`.
👍

**Comment by dconnolly** [comment]
@teor2345 do a rebase onto ZcashFoundation/zebra#main to get the CI check fixes. 🙏

**Comment by teor2345** [comment]
I've done a basic stub that unconditionally adds blocks to `ZebraState`.

I had some trouble getting the futures and types to work, so I just returned `Ok(())`. I'd love some help with complex Rust type handling here. (The compiler errors were starting to look a lot like C++.)

**Comment by teor2345** [comment]
(Also, I didn't write any tests. I should be ok with writing tests, once the future and type issues are sorted out.)

**Comment by yaahc** [comment]
> @teor2345 do a rebase onto ZcashFoundation/zebra#main to get the CI check fixes. pray

In the future I think it would be better to just do a merge with a merge commit with the expectation that it will get squashed + rebased when the PR eventually goes in. You rebasing it makes it so I cannot select the "see changes since your last review".

It's not a big deal, because I can still see the date that each commit was made so its easy to see which ones are new, but still I think its a good idea to just avoid rebasing unless its for a branch that isn't going to be squashed later. That said, if you've been making these commits with the intent of not squashing them then feel free to ignore this advice.

File: zebra-consensus/src/verify.rs
Comment (yaahc) [inline zebra-consensus/src/verify.rs]: NIT: I think it might make more sense to call this `ZebraVerify`, not that it matters though given that it isn't exported.

File: zebra-consensus/src/verify.rs
Comment (yaahc) [inline zebra-consensus/src/verify.rs]: Is there something preventing us from making the response type the `BlockHeaderHash` from the start with the stub?

File: zebra-consensus/src/verify.rs
Comment (yaahc) [inline zebra-consensus/src/verify.rs]:
```suggestion
// TODO(jane): Error = Report ?
```

heh, I'll handle this one, I am on the hook for the rest of the `Report` stuff for services too. The reason we're still using `Box<dyn Error...>` here and elsewhere is because `tower` hasn't implemented `map_err` for `Service` yet, so we cannot convert the return type of the Service for `Buffer` easily. We might just implement a `map_err`-like service specifically for mapping to a `Report` to do this short term but that's all stuff on my plate so yea.

File: zebra-consensus/src/verify.rs
Comment (yaahc) [inline zebra-consensus/src/verify.rs]:
```suggestion
type Error = Box<dyn Error + Send + Sync + 'static>;
```

File: zebra-consensus/src/verify.rs
Comment (yaahc) [inline zebra-consensus/src/verify.rs]:
```suggestion
type Response = ();
```

File: zebra-consensus/src/verify.rs
Comment (yaahc) [inline zebra-consensus/src/verify.rs]: doesn't look like there is any error handling to be done here, if we end up with functions that do return errors being called here just make sure they're inside the `async` block which has a return type of `Result` and it will be all good.

**Review by yaahc — CHANGES_REQUESTED** [review]
Some minor changes requested, also the various clippy lints that are being triggered. If you haven't already it might be good to setup `rust-analyzer` to default to running `cargo clippy` instead of `cargo check`

File: design/design.md
Comment (hdevalence) [inline design/design.md]: I'm not sure how we should describe it in the `design.md`, but one thing that will likely end up happening is that rather than exporting a concrete type that implements `Service`, we would export a function that returns `impl Service<...>` with some generic bounds, as is done in `zebra-network`.

The reason is that we want to be able to make use of Tower layers (e.g., using a `Buffer` to make the service `Clone`, or adding metrics,  or a retry policy, or timeouts, or...), and those work by wrapping an inner service type in another wrapper type that also implements `Service`.  Returning `impl Service<...>` means that we are free to return a `Buffer<ZebraState>` or a `Buffer<Retry<Timeout<PeakEwma<...>>>>` or whatever combination of layers we want.

File: zebra-consensus/src/verify.rs
Comment (hdevalence) [reply zebra-consensus/src/verify.rs]: Maybe `BlockVerifier`?

Comment (hdevalence) [reply zebra-consensus/src/verify.rs]: Then we can have `TransactionVerifier`, `ScriptVerifier`, etc.

**Comment by teor2345** [comment]
> In the future I think it would be better to just do a merge with a merge commit with the expectation that it will get squashed + rebased when the PR eventually goes in. You rebasing it makes it so I cannot select the "see changes since your last review".
> 
> It's not a big deal, because I can still see the date that each commit was made so its easy to see which ones are new, but still I think its a good idea to just avoid rebasing unless its for a branch that isn't going to be squashed later. That said, if you've been making these commits with the intent of not squashing them then feel free to ignore this advice.

I agree, "no rebases" is also how I like to review.

We always do a final rebase before merges, so I don't think it matters whether I squash or not. (It's the rebase that removes the merge commits.)

(For most PRs, GitHub does a hidden automatic merge, then runs it in CI. So I may not even have needed to rebase here.)

**Comment by teor2345** [comment]
> If you haven't already it might be good to setup `rust-analyzer` to default to running `cargo clippy` instead of `cargo check`

Ah, that's what that emacs lsp-rust-analyzer setting is for:
`(setq lsp-rust-analyzer-cargo-watch-command "clippy")`

**Comment by dconnolly** [comment]
> > @teor2345 do a rebase onto ZcashFoundation/zebra#main to get the CI check fixes. pray
> 
> In the future I think it would be better to just do a merge with a merge commit with the expectation that it will get squashed + rebased when the PR eventually goes in. You rebasing it makes it so I cannot select the "see changes since your last review".
> 
> It's not a big deal, because I can still see the date that each commit was made so its easy to see which ones are new, but still I think its a good idea to just avoid rebasing unless its for a branch that isn't going to be squashed later. That said, if you've been making these commits with the intent of not squashing them then feel free to ignore this advice.

We don't enforce squash so we'd have to be careful with that

File: zebra-consensus/src/verify.rs
Comment (teor2345) [reply zebra-consensus/src/verify.rs]: Fixed in dec5c37fb5bc08eab967064ca02069848ec117ff to 12c20d72d11eb58564afc61c34f9c83794950853.

File: zebra-consensus/src/verify.rs
Comment (teor2345) [reply zebra-consensus/src/verify.rs]: Switched the TODO to you in c620e4b42dc367ffd6fd8b20bfeade182eab7ff2.

File: zebra-consensus/src/verify.rs
Comment (teor2345) [reply zebra-consensus/src/verify.rs]: Fixed in c620e4b42dc367ffd6fd8b20bfeade182eab7ff2.

File: zebra-consensus/src/verify.rs
Comment (teor2345) [reply zebra-consensus/src/verify.rs]: Fixed in c620e4b42dc367ffd6fd8b20bfeade182eab7ff2.

File: zebra-consensus/src/verify.rs
Comment (teor2345) [reply zebra-consensus/src/verify.rs]: Fixed in c620e4b42dc367ffd6fd8b20bfeade182eab7ff2, with a comment explaining that zebra_state already does all the error handling we need.

File: design/design.md
Comment (teor2345) [reply design/design.md]: Fixed in dec5c37fb5bc08eab967064ca02069848ec117ff so that it's similar to the zebra-network service docs.

**Comment by teor2345** [comment]
> If you haven't already it might be good to setup `rust-analyzer` to default to running `cargo clippy` instead of `cargo check`

`cargo clippy` has no errors as of c620e4b42dc367ffd6fd8b20bfeade182eab7ff2.

File: zebra-consensus/src/verify.rs
Comment (teor2345) [reply zebra-consensus/src/verify.rs]: I was struggling with Rust lifetimes. It would be really helpful if rustc could give better hints about taking references for `into()`, and using `async move` when necessary.

Fixed in 7f86eae3f0f78b23dc7ca8a48e9e6509a8dba17f.

**Comment by teor2345** [comment]
I think the code works now, but I still have to:
* write tests
* squash the fixups and 07ef270e45242b30ff30f1d987692853912b74d7

In my next PR, I'll split blocks into headers and transactions, and make stubs for those services. Then I can actually start verifying.

File: zebra-consensus/src/verify.rs
Comment (yaahc) [inline zebra-consensus/src/verify.rs]: It's not that we're ignoring errors, its just that there are none.

File: zebra-consensus/src/verify.rs
Comment (yaahc) [reply zebra-consensus/src/verify.rs]: I think this one might have been missed

File: zebra-consensus/src/verify.rs
Comment (yaahc) [inline zebra-consensus/src/verify.rs]: I think this one needs to be long lived, we should store the service that this returns in the `BlockVerifier` and then call `in_memory::init` inside of `BlockVerifier::init` when constructing the block verifier service.

File: zebra-consensus/src/verify.rs
Comment (yaahc) [inline zebra-consensus/src/verify.rs]: This should also just reference the service in `self`

**Review by yaahc — CHANGES_REQUESTED** [review]

File: zebra-consensus/src/verify.rs
Comment (teor2345) [reply zebra-consensus/src/verify.rs]: I think I might be confused by the design here. (Or the terminology.)

ZebraState rejects blocks if another block is added at the same height. (The block can be the same as the existing block, or a chain fork.)
https://github.com/ZcashFoundation/zebra/blob/main/zebra-state/src/in_memory/block_index.rs#L31

If we make BlockVerifier satisfy the ZebraState preconditions, then there will be no errors here. But I haven't done that yet.

How should BlockVerifier handle `Err("forks in the chain aren't supported yet")` ?

Comment (yaahc) [reply zebra-consensus/src/verify.rs]: oh, my bad I misunderstood what this was doing. I didn't realize that you're just discarding the result, yea error handling will be necessary, we will probably have to switch to using real error types so we can handle them effectively.

Comment (teor2345) [reply zebra-consensus/src/verify.rs]: Ok, I'll mark that down as your TODO?

Comment (teor2345) [reply zebra-consensus/src/verify.rs]: I split the TODOs between us both in 74b6cc790379fa50d7bf24a238aa5b556a3d6f9d.

File: zebra-consensus/src/verify.rs
Comment (teor2345) [reply zebra-consensus/src/verify.rs]: I'm struggling with the types and lifetimes in this change.

Commit 30b1c848046b0c499c2af60ee577173459039eba doesn't have this change, it compiles fine.

Commit d47e77b02c70ae9454d708159e0710b8ac60c141 has my attempt at this change, with `error: the `call` method cannot be invoked on a trait object`. The full error is in the commit message.

Comment (teor2345) [reply zebra-consensus/src/verify.rs]: CI also shows an extra error: "the `poll_ready` method cannot be invoked on a trait object"
https://github.com/ZcashFoundation/zebra/pull/448/checks?check_run_id=757541268#step:5:759

File: zebra-consensus/src/verify.rs
Comment (yaahc) [inline zebra-consensus/src/verify.rs]: In order to store the inner service behind a static reference we'd either need the zebra state service to qualify for const promotion or we'd need to allocate it in a static using internal mutability such as with a `once_cell` or a `lazy_static`. I don't think it can qualify for const promotion because the state service has multiple maps which use dyn allocations and what not and I don't think we can initialize it in a static because the types are existential so we cannot name the type for the static, which is an issue because static items don't support type inference.

also I think `&'static mut` references are unsafe

for more details on const promotion (apparently also called static promotion) check out these links:

* https://github.com/rust-lang/rfcs/blob/master/text/1414-rvalue_static_promotion.md
* https://github.com/rust-lang/const-eval/blob/master/promotion.md

So there are two approaches we can take here. We can either Box the service and make the associated future be generic, or we can make them both generic. I went with the latter option when I was refactoring `connect.rs` yesterday, but I think boxing the service might result in less verbose code.

Comment (hdevalence) [reply zebra-consensus/src/verify.rs]: Another note is that the verification service's constructor needs to take the state service as a parameter, rather than constructing it internally, because the state system should be pluggable.  This means that the ctor needs to either take a boxed service, or the verification service needs to be generic over `S` where `S: Service<zebra_storage::Request, Response = zebra_storage::Response, Error = Box<dyn Error + Send + Sync + 'static>`.

File: zebra-consensus/src/verify.rs
Comment (yaahc) [inline zebra-consensus/src/verify.rs]: this looks perfect

File: zebra-consensus/src/verify.rs
Comment (yaahc) [inline zebra-consensus/src/verify.rs]: if we had .map for services we could box the future to get rid of the generic while still having the type be sized.

**Review by yaahc — DISMISSED** [review]
🛳️

File: zebra-consensus/src/verify.rs
Comment (teor2345) [reply zebra-consensus/src/verify.rs]: Great! I'll squash the branch, force-push, and add tests. Then we can do a final review.

File: zebra-consensus/src/verify.rs
Comment (teor2345) [reply zebra-consensus/src/verify.rs]: Done in d19fd2ec504aa4fc813338ae4e0b5f1dbb169294.

File: zebra-consensus/src/verify.rs
Comment (teor2345) [reply zebra-consensus/src/verify.rs]: Done in d19fd2ec504aa4fc813338ae4e0b5f1dbb169294 and previous commits.

File: zebra-consensus/src/verify.rs
Comment (teor2345) [reply zebra-consensus/src/verify.rs]: Fixed in d19fd2ec504aa4fc813338ae4e0b5f1dbb169294.

**Comment by teor2345** [comment]
I've just pushed the squashed version of the branch, I still need to add tests.

**Comment by teor2345** [comment]
I've added a basic verify test, which passes.

I also tried to add a round-trip test that accesses the inner state, but it's failing due to some internal service or future error. If it's something simple, let's fix it and merge.

But if it's going to take a while, let's merge the working test, and open another PR to sort out the round-trip test.

File: zebra-consensus/Cargo.toml
Comment (yaahc) [inline zebra-consensus/Cargo.toml:21]: I think this _might_ have negative interactions with the features that are enabled when not using dev-dependencies. AFAIK this bug in cargo was fixed in nightly recently so its probably fine but we might want to open an issue to verify that we're not pulling in more of tokio than wanted when compiling just `zebra-concensus` without tests

**Review by yaahc — DISMISSED** [review]

**Comment by hdevalence** [comment]
@teor2345 I think I found the problem causing the test to fail; the state update future was never driven to completion.  If you want to cherry-pick it, I pushed a commit that fixes it to the `verify-stubs` branch in this repo (since I don't think I can edit the branch in your fork).

**Comment by hdevalence** [comment]
This looks great.  The only remaining comment I have is that I think that the `where` bounds would be a lot simpler if they were written in terms of the state service rather than the service's future (since the future is already constrained by the `Service` trait).  I checked by changing them this way locally, but since working with Tower bounds + Rust errors can be unintuitive at first, I'd be happy to go through the process of changing all the bounds if you think it would be helpful.

File: zebra-consensus/Cargo.toml
Comment (teor2345) [reply zebra-consensus/Cargo.toml:21]: I used GitHub's "reference in new issue" to open #468.

**Comment by teor2345** [comment]
I think we're done with these stubs.

Here's what I did:
* summarised the commit message for my "failing test" commit,
* cherry-picked @hdevalence's commits,
* pushed all the header hashing down into `zebra_state::in_memory::block_index`.
(It's a cleaner design, and I think it will support chain reorganisations better.)

I'm ready for final review, and a rebase merge.

**Review by yaahc — APPROVED** [review]
LGTM

**Review by hdevalence — APPROVED** [review]
LGTM

**Pull request merged**
